llvm-bazel/zlib.bzl

```diff
@@ -0,0 +1,113 @@
+
+Most users should pick one of the explicit rules to configure their use of zlib
+with LLVM:
+- `llvm_zlib_external` will link against an external Bazel zlib repository.
+- `llvm_zlib_system` will link against the system zlib (non-hermetically).
+- 'llvm_zlib_disable` will disable zlib completely.
+
+If you would like to make your build configurable, you can use
+`llvm_zlib_from_env`. By default, this will disable zlib, but will inspect
+the environment variable (most easily set with a `--repo_env` flag to the
+Bazel invocation) `BAZEL_LLVM_ZLIB_STRATEGY`. If it is set to `external`,
+then it will behave the same as `llvm_zlib_external`. If it is set to
+`system` then it will behave the same as `llvm_zlib_system`. Any other
+setting will disable zlib the same as not setting it at all.
+"""
+
+def _llvm_zlib_external_impl(repository_ctx):
+    repository_ctx.template(
+        "BUILD",
+        repository_ctx.attr._external_build_template,
+        substitutions = {
+            "@external_zlib_repo//:zlib_rule": str(repository_ctx.attr.external_zlib),
+        },
+        executable = False,
+    )
+
+llvm_zlib_external = repository_rule(
+    implementation = _llvm_zlib_external_impl,
+    local = True,
+    attrs = {
+        "_external_build_template": attr.label(
+            default = Label("//:zlib_external.BUILD"),
+            allow_single_file = True,
+        ),
+        "external_zlib": attr.label(
+            doc = "The dependency that should be used for the external zlib library.",
+            mandatory = True,
+        ),
+    },
+)
+
+def _llvm_zlib_system_impl(repository_ctx):
+    repository_ctx.template(
+        "BUILD",
+        repository_ctx.attr._system_build_template,
+        executable = False,
+    )
+
+llvm_zlib_system = repository_rule(
+    implementation = _llvm_zlib_system_impl,
+    local = True,
+    attrs = {
+        "_system_build_template": attr.label(
+            default = Label("//:zlib_system.BUILD"),
+            allow_single_file = True,
+        ),
+    },
+)
+
+def _llvm_zlib_disable_impl(repository_ctx):
+    repository_ctx.template(
+        "BUILD",
+        repository_ctx.attr._disable_build_template,
+        executable = False,
+    )
+
+llvm_zlib_disable = repository_rule(
+    implementation = _llvm_zlib_disable_impl,
+    local = True,
+    attrs = {
+        "_disable_build_template": attr.label(
+            default = Label("//:zlib_disable.BUILD"),
+            allow_single_file = True,
+        ),
+    },
+)
+
+def _llvm_zlib_from_env_impl(repository_ctx):
+    zlib_strategy = repository_ctx.os.environ.get("BAZEL_LLVM_ZLIB_STRATEGY")
+    if zlib_strategy == "external":
+        _llvm_zlib_external_impl(repository_ctx)
+    elif zlib_strategy == "system":
+        _llvm_zlib_system_impl(repository_ctx)
+    else:
+        _llvm_zlib_disable_impl(repository_ctx)
+
+llvm_zlib_from_env = repository_rule(
+    implementation = _llvm_zlib_from_env_impl,
+    local = True,
```

Comment:
Declare `BAZEL_LLVM_ZLIB_STRATEGY` in `environ` and then I think this can be not local as well.